smtp.go

```diff
@@ -2,209 +2,334 @@ package maddy
-		tokens, ok := tokens[name]
-		if !ok {
-			continue
-		}
-
-		if be != nil {
-			return nil, errors.New("multiple SMTP upstreams specified")
-		}
+func (u SMTPUser) Logout() error {
+	return nil
+}
 
-		var err error
-		be, err = newUpstream(caddyfile.NewDispenserTokens("", tokens))
-		if err != nil {
-			return nil, fmt.Errorf("failed to create `%v` upstream: %v", name, err)
-		}
-	}
+type SMTPEndpoint struct {
+	Auth      module.AuthProvider
+	serv      *smtp.Server
+	name      string
+	domain    string
+	listeners []net.Listener
+	pipeline  []SMTPPipelineStep
+}
 
-	if be == nil {
-		return nil, errors.New("missing SMTP upstream")
-	}
+func (endp *SMTPEndpoint) Name() string {
+	return "smtp"
+}
 
-	return be, nil
+func (endp *SMTPEndpoint) InstanceName() string {
+	return endp.name
 }
 
-func newSMTPProxy(d caddyfile.Dispenser) (smtp.Backend, error) {
-	for d.Next() {
-		args := d.RemainingArgs()
+func (endp *SMTPEndpoint) Version() string {
+	return VersionStr
+}
 
-		if len(args) == 1 {
-			addr, err := standardizeAddress(args[0])
+func NewSMTPEndpoint(instName string, cfg config.CfgTreeNode) (module.Module, error) {
+	endp := new(SMTPEndpoint)
+	endp.name = instName
+	var (
+		err     error
+		tlsConf *tls.Config
+		tlsSet  bool
+
+		localDeliveryDefault  string
+		localDeliveryOpts     map[string]string
+		remoteDeliveryDefault string
+		remoteDeliveryOpts    map[string]string
+	)
+	maxIdle := -1
+	maxMsgBytes := -1
+
+	insecureAuth := false
+	ioDebug := false
+	for _, entry := range cfg.Childrens {
+		switch entry.Name {
+		case "auth":
+			endp.Auth, err = authProvider(entry.Args)
+			if err != nil {
+				return nil, err
+			}
+		case "hostname":
+			if len(entry.Args) != 1 {
+				return nil, errors.New("hostname: expected 1 argument")
+			}
+			endp.domain = entry.Args[0]
+		case "maxidle":
+			if len(entry.Args) != 1 {
+				return nil, errors.New("maxidle: expected 1 argument")
+			}
+			maxIdle, err = strconv.Atoi(entry.Args[0])
+			if err != nil {
+				return nil, errors.New("maxidle: invalid integer value")
+			}
+		case "maxmsgbytes":
+			if len(entry.Args) != 1 {
+				return nil, errors.New("maxmsgbytes: expected 1 argument")
+			}
+			maxMsgBytes, err = strconv.Atoi(entry.Args[0])
 			if err != nil {
+				return nil, errors.New("maxmsgbytes: invalid integer value")
+			}
+		case "tls":
+			tlsConf = new(tls.Config)
+			if err := setTLS(entry.Args, &tlsConf); err != nil {
 				return nil, err
 			}
+			tlsSet = true
+			if tlsConf == nil {
+				log.Printf("smtp %s: TLS is disabled, this is insecure configuration and should be used only for testing!", instName)
+				insecureAuth = true
+			}
+		case "insecureauth":
+			log.Printf("smtp %s: authentication over unencrypted connections is allowed, this is insecure configuration and should be used only for testing!", instName)
+			insecureAuth = true
+		case "iodebug":
+			log.Printf("smtp %v: I/O debugging is on!\n", instName)
+			ioDebug = true
+		case "local-delivery":
```

Comment:
We need to decide what conventions we use for these. Caddy uses underscores, so that would give things like `insecure_auth` and `local_delivery`.

Reply:
Will change to underscores.